syntax = "proto3";

service FileService {
    rpc getFile(GetFileRequest) returns (GetFileResponse) {}
    rpc saveFile(SaveFileRequest) returns (SaveFileResponse) {}
    rpc cloneFile(CloneFileRequest) returns (CloneFileResponse) {}
    rpc ReplicateFile(ReplicateFileRequest) returns (ReplicateFileResponse) {}
}

// GET FILE
message GetFileRequest {
    string file_name = 1;
}

message GetFileResponse {
    string file_name = 1;
    int32 total_parts = 2;
    repeated Parts parts = 3;
}

message Parts {
    int32 part_id = 1;
    string part_name = 2;
    bytes data = 3;
}

// SAVE FILE

message SaveFileRequest {
    string global_name = 1;
    string file_name = 2;
    bytes content = 3;
}

message SaveFileResponse {
    bool status = 1;
    string file_name = 2;
}

// Mensaje para solicitar la replica
message ReplicateFileRequest {
  repeated string file_parts = 1;  
  string source_directory = 2;     
  string target_directory = 3;  
}

// Mensaje para responder a una solicitud de replica
message ReplicateFileResponse {
  bool success = 1;
}



// CLONE ORDER

message CloneFileRequest {
    string file_name = 1;
    string recieve_ip = 2;
    string recieve_port = 3;
}

message CloneFileResponse{
    bool status = 1;
    string file_name = 2;
}